### PR TITLE
Add support for policy based span sampling

### PR DESCRIPTION
**Pull request by xshipeng**

### CODE REVIEW

File: proxy/src/main/java/com/wavefront/agent/sampler/SpanSampler.java [hunk 1]
```diff
@@ -27,18 +39,45 @@
  * @author Han Zhang (zhanghan@vmware.com)
  */
 public class SpanSampler {
+  private static final int EXPIRE_AFTER_ACCESS_SECONDS = 600;
+  private static final String SPAN_SAMPLING_POLICY_TAG = "_sampledByPolicy";
```
[the remark this answers is not in the document]
Comment (xshipeng) [reply proxy/src/main/java/com/wavefront/agent/sampler/SpanSampler.java]: I talked with Han offline and we decided to make SPAN_SAMPLING_POLICY_TAG public in SpanSampler for  SpanHandlerImpl to use.

File: proxy/src/main/java/com/wavefront/agent/sampler/SpanSampler.java [hunk 2]
```diff
@@ -27,18 +39,45 @@
...
   private final boolean alwaysSampleErrors;
+  private final LoadingCache<String, Predicate<Span>> spanPredicateCache = Caffeine.newBuilder().expireAfterAccess(EXPIRE_AFTER_ACCESS_SECONDS,
+      TimeUnit.SECONDS).build(new CacheLoader<String, Predicate<Span>>() {
+    @Override
+    @Nullable
+    public Predicate<Span> load(@NonNull String key) {
+      try {
+        return Predicates.fromPredicateEvalExpression(key);
+      } catch (ExpressionSyntaxException ex) {
+        logger.severe("Policy expression " + key + " is invalid: " + ex.getMessage());
```
[the remark this answers is not in the document]
Comment (xshipeng) [reply proxy/src/main/java/com/wavefront/agent/sampler/SpanSampler.java:53]: We have the syntax exception checking when the user updates/creates the policy, so this exception should not be expected to happen in the proxy.

File: proxy/src/main/java/com/wavefront/agent/sampler/SpanSampler.java [hunk 3]
```diff
@@ -63,6 +102,29 @@ public boolean sample(Span span, @Nullable Counter discarded) {
     if (isForceSampled(span)) {
       return true;
     }
+    // Policy based span sampling
```
[the remark this answers is not in the document]
Comment (xshipeng) [reply proxy/src/main/java/com/wavefront/agent/sampler/SpanSampler.java:94]: I talked with Han offline and we decided to remove the force sampling for spans with error tag and sampling.priority tag.

**Review by hanwavefront — COMMENTED** [review]
Make sure to remove `traceAlwaysSampleErrors` everywhere (e.g., `wavefront.conf.default`). Otherwise LGTM.

**Review by hanwavefront — APPROVED** [review]

**Pull request merged**